streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStreamImpl.java

```diff
@@ -0,0 +1,196 @@
+                return aggregate + 1;
+            }
+        }, windows, Serdes.Long());
+    }
+
+
+    private <W extends Window, T> StateStoreSupplier windowedStore(final Serde<T> aggValSerde,
+                                                                   final Windows<W> windows) {
+        return storeFactory(aggValSerde, windows.name())
+            .windowed(windows.maintainMs(), windows.segments, false)
+            .build();
+
+    }
+
+    private <T> Stores.PersistentKeyValueFactory<K, T> storeFactory(final Serde<T> aggValueSerde,
+                                                                    final String name) {
+        return Stores.create(name)
+            .withKeys(keySerde)
+            .withValues(aggValueSerde)
+            .persistent();
+
+    }
+
+    private <T> KTable<K, T> doAggregate(
+        final KStreamAggProcessorSupplier<K, ?, V, T> aggregateSupplier,
+        final String functionName,
+        final StateStoreSupplier storeSupplier) {
+
+        final String aggFunctionName = topology.newName(functionName);
+
+        final String sourceName = repartitionIfRequired();
+
+        topology.addProcessor(aggFunctionName, aggregateSupplier, sourceName);
+        topology.addStateStore(storeSupplier, aggFunctionName);
+
+        return new KTableImpl<>(topology,
+                                aggFunctionName,
+                                aggregateSupplier,
+                                sourceName.equals(this.name) ? sourceNodes
+                                                             : Collections.singleton(sourceName));
+    }
+
+    /**
+     * If repartitioning is required then add a new co-partitioned internal topic and configure the
+     * source & sink.
+     *
+     * @return the new sourceName if repartitioned. Otherwise the name of this stream
+     */
+    private String repartitionIfRequired() {
+        if (!repartitionRequired) {
+            return this.name;
+        }
+
+        final Serializer<K> keySerializer = keySerde == null ? null : keySerde.serializer();
+        final Deserializer<K> keyDeserializer = keySerde == null ? null : keySerde.deserializer();
+        final Serializer<V> valueSerializer = valSerde == null ? null : valSerde.serializer();
+        final Deserializer<V> valueDeserializer = valSerde == null ? null : valSerde.deserializer();
+
+        final String sinkName = topology.newName(KStreamImpl.SINK_NAME);
+        final String topic = this.name + KStreamImpl.REPARTITION_TOPIC_SUFFIX;
+        final Set<String> allSourceNodes = new HashSet<>(sourceNodes);
+        final String sourceName = topology.newName(KStreamImpl.SOURCE_NAME);
+        allSourceNodes.add(sourceName);
+        topology.addInternalTopic(topic);
+        topology.addSink(sinkName, topic, keySerializer, valueSerializer, this.name);
+        topology.addSource(sourceName, keyDeserializer, valueDeserializer, topic);
+        topology.copartitionSources(allSourceNodes);
```

Comment:
Strictly speaking we don't, but it is my understanding that if it is not co-partitioned the new topic will be created with the default number of partitions, which could be completely wrong for this topic. My assumption is that you would want the internal topic to have the same number of partitions as the source topic in this case.